csharp/src/Google.Protobuf.Test/Reflection/CustomOptionsTest.cs

```diff
@@ -0,0 +1,146 @@
+            Assert.True(options.TryGetInt32(1, out intValue));
+            Assert.AreEqual(1234567, intValue);
+
+            string stringValue;
+            // No ByteString stored values
+            Assert.False(options.TryGetString(1, out stringValue));
+            // Nothing stored for field 2
+            Assert.False(options.TryGetInt32(2, out intValue));
+        }
+
+        [Test]
+        public void SimpleStringTest()
+        {
+            var stream = new MemoryStream();
+            var output = new CodedOutputStream(stream);
+            output.WriteTag(MakeTag(1, WireType.LengthDelimited));
+            output.WriteString("value");
+            output.Flush();
+            stream.Position = 0;
+            var input = new CodedInputStream(stream);
+            input.ReadTag();
+
+            var options = CustomOptions.Empty;
+            options = options.ReadOrSkipUnknownField(input);
+
+            string stringValue;
+            Assert.True(options.TryGetString(1, out stringValue));
+            Assert.AreEqual("value", stringValue);
+
+            int intValue;
+            // No numeric stored values
+            Assert.False(options.TryGetInt32(1, out intValue));
+            // Nothing stored for field 2
+            Assert.False(options.TryGetString(2, out stringValue));
+        }
+
+        [Test]
+        public void ScalarOptions()
+        {
+            var options = CustomOptionOtherValues.Descriptor.CustomOptions;
+            AssertOption(-100, options.TryGetInt32, 7705709);
+            AssertOption(12.3456789f, options.TryGetSingle, 7675390);
+            AssertOption(1.234567890123456789d, options.TryGetDouble, 7673293);
+            AssertOption("Hello, \"World\"", options.TryGetString, 7673285);
+            AssertOption(ByteString.CopyFromUtf8("Hello\0World"), options.TryGetBytes, 7673238);
+            AssertOption((int) TestEnumType.TestOptionEnumType2, options.TryGetInt32, 7673233);
+        }
+
+        [Test]
+        public void MessageOptions()
+        {
+            var options = VariousComplexOptions.Descriptor.CustomOptions;
+            AssertOption(new ComplexOptionType1 { Foo = 42, Foo4 = { 99, 88 } }, options.TryGetMessage, 7646756);
+            AssertOption(new ComplexOptionType2
+                {
+                    Baz = 987, Bar = new ComplexOptionType1 { Foo = 743 },
+                    Fred = new ComplexOptionType4 { Waldo = 321 },
+                    Barney = { new ComplexOptionType4 { Waldo = 101 }, new ComplexOptionType4 { Waldo = 212 } }
+                },
+                options.TryGetMessage, 7636949);
+            AssertOption(new ComplexOptionType3 { Qux = 9 }, options.TryGetMessage, 7636463);
+        }
+
+        private void AssertOption<T>(T expected, OptionFetcher<T> fetcher, int field)
+        {
+            T actual;
+            Assert.IsTrue(fetcher(field, out actual));
+            Assert.AreEqual(expected, actual);
+        }
+    }
+}
```

Comment:
Add more tests which has defined in the proto file? Like TestMessageWithCustomOptions etc.